aten/src/ATen/native/Distributions.cpp

```diff
@@ -126,71 +126,41 @@ DEFINE_DISPATCH(random_stub);
-  return at::empty_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT).bernoulli_(self, gen);
+  Tensor result = at::empty_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+  result.bernoulli_(self, gen);
+  return result;
 }
 
 Tensor bernoulli(const Tensor& self, double p, c10::optional<Generator> gen) {
-  return at::empty_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT).bernoulli_(p, gen);
+  Tensor result = at::empty_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+  result.bernoulli_(p, gen);
+  return result;
 }
 
 Tensor& bernoulli_out(Tensor& result, const Tensor& self, c10::optional<Generator> gen) {
-  // result.resize_as_(self) requires self to have same dtype as result, so we
-  // use resize_ instead.
-  // TODO: Fix resize_as_. See pytorch/pytorch#11665.
-  result.resize_(self.sizes()).bernoulli_(self, gen);
-  namedinference::propagate_names(result, self);
-  return result;
+  return at::native::templates::bernoulli_out_impl<BernoulliStub, Generator>(result, self, gen);
 }
 
-Tensor& bernoulli_tensor_cpu_(Tensor& self, const Tensor& p_, c10::optional<Generator> gen) {
-  NoNamesGuard guard;
-  AT_DISPATCH_ALL_TYPES_AND(at::ScalarType::Bool, self.scalar_type(), "bernoulli_tensor_cpu_self_", [&] {
-    CPUGeneratorImpl* generator = get_generator_or_default<CPUGeneratorImpl>(gen, detail::getDefaultCPUGenerator());
-    // See Note [Acquire lock when using random generators]
-    std::lock_guard<std::mutex> lock(generator->mutex_);
-    using self_t = scalar_t;
-    if (p_.scalar_type() == kDouble) {
-      auto p = std::get<0>(expand_inplace(self, p_.to(kCPU)));
-      CPU_tensor_apply2<self_t, double>(
-        self, p, [generator](self_t& ret_val, double& p_val) {
-          at::bernoulli_distribution<double> bernoulli(p_val);
-          ret_val = static_cast<self_t>(bernoulli(generator));
-        });
-    } else {
-      AT_DISPATCH_FLOATING_TYPES(p_.scalar_type(), "bernoulli_tensor_cpu_p_", [&] {
-        auto p = std::get<0>(expand_inplace(self, p_.to(kCPU)));
-        using p_t = scalar_t;
-        CPU_tensor_apply2<self_t, p_t>(
-          self, p, [generator](self_t& ret_val, p_t& p_val) {
-            at::bernoulli_distribution<float> bernoulli(p_val);
-            ret_val = static_cast<self_t>(bernoulli(generator));
-        });
-      });
-    }
-  });
-  return self;
+Tensor& bernoulli_(Tensor& self, const Tensor& p_, c10::optional<Generator> gen) {
+  return at::native::templates::bernoulli_impl_<BernoulliStub, Generator>(self, p_, gen);
 }
 
-Tensor& bernoulli_scalar_cpu_(Tensor& self, double p, c10::optional<Generator> gen) {
-  TORCH_CHECK(0 <= p && p <= 1, "bernoulli_ expects p to be in [0, 1], but got p=", p);
-#if AT_MKL_ENABLED()
-  if (cpuinfo_initialize() && cpuinfo_vendor_intel == cpuinfo_get_processor(0)->core->vendor) {
-    bernoulli_mkl_stub(kCPU, self, p, gen);
-    return self;
-  }
-#endif
-  AT_DISPATCH_ALL_TYPES_AND(at::ScalarType::Bool, self.scalar_type(), "bernoulli_scalar_cpu_", [&] {
```

Comment:
The old code here makes AMD can move to using the default version even though `AT_MKL_ENABLED` is `True`, and it guaranteed the `AT_ASSERTM` will never be reached.